File: Golang/Medium/1722_minimize_hamming_distance_after_swap_operations.go
package main

import "fmt"

// Solution groups all approaches for 1722. Minimize Hamming Distance After
// Swap Operations.
//
// Indices connected through allowedSwaps form components within which source
// elements can be freely rearranged. The minimum Hamming distance equals the
// number of target positions that cannot be matched by any available source
// value within the same component.
type Solution struct{}

// minimumHammingDistance returns the minimum Hamming distance using Union-Find.
//
// Unions all swap pairs into components, groups source values by component root
// into a frequency map, then greedily matches target values against the map.
//
// Time complexity:  O(n + m·α(n)) where m = len(allowedSwaps).
// Space complexity: O(n)
func (sol Solution) minimumHammingDistance(source, target []int, allowedSwaps [][]int) int {
	return 0 // TODO
}

// minimumHammingDistanceDFS returns the minimum Hamming distance using DFS to
// find connected components on the swap graph.
//
// Builds an adjacency list from allowedSwaps, finds each connected component
// via DFS, then counts how many target values in the component are unmatched
// by source values using frequency maps and min(src_freq, tgt_freq).
//
// Time complexity:  O(n + m) where m = len(allowedSwaps).
// Space complexity: O(n + m)
func (sol Solution) minimumHammingDistanceDFS(source, target []int, allowedSwaps [][]int) int {
	return 0 // TODO
}

func main() {
	testCases := []struct {
		source       []int
		target       []int
		allowedSwaps [][]int
		expected     int
	}{
		{
			source: []int{1, 2, 3, 4}, target: []int{2, 1, 4, 5},
			allowedSwaps: [][]int{{0, 1}, {2, 3}}, expected: 1,
		},
		{
			source: []int{1, 2, 3, 4}, target: []int{1, 3, 2, 4},
			allowedSwaps: [][]int{}, expected: 2,
		},
		{
			source: []int{5, 1, 2, 4, 3}, target: []int{1, 5, 4, 2, 3},
			allowedSwaps: [][]int{{0, 4}, {4, 2}, {1, 3}, {1, 4}}, expected: 0,
		},
		{source: []int{1}, target: []int{1}, allowedSwaps: [][]int{}, expected: 0},
		{source: []int{1}, target: []int{2}, allowedSwaps: [][]int{}, expected: 1},
	}

	sol := Solution{}

	for _, approach := range []struct {
		label string
		fn    func([]int, []int, [][]int) int
	}{
		{"minimumHammingDistance   ", sol.minimumHammingDistance},
		{"minimumHammingDistanceDFS", sol.minimumHammingDistanceDFS},
	} {
		fmt.Printf("--- %s ---\n", approach.label)
		for i, tc := range testCases {
			result := approach.fn(tc.source, tc.target, tc.allowedSwaps)
			status := "PASS"
			if result != tc.expected {
				status = "FAIL"
			}
			fmt.Printf("  Example %d | source=%v target=%v | Expected: %d | Result: %d | [%s]\n",
				i+1, tc.source, tc.target, tc.expected, result, status)
		}
	}
}

File: Python/Medium/1722_minimize_hamming_distance_after_swap_operations.py
"""
1722. Minimize Hamming Distance After Swap Operations

Problem Statement:
You are given two integer arrays source and target, both of length n, and an
array allowedSwaps where allowedSwaps[i] = [ai, bi] means you can swap the
elements at indices ai and bi of source any number of times.

Return the minimum Hamming distance of source and target after performing any
amount of swap operations on source.

Constraints:
- n == source.length == target.length
- 1 <= n <= 10^5
- 1 <= source[i], target[i] <= 10^5
- 0 <= allowedSwaps.length <= 10^5
- allowedSwaps[i].length == 2
- 0 <= ai, bi <= n - 1
- ai != bi

Examples:

Example 1:
    Input: source = [1,2,3,4], target = [2,1,4,5], allowedSwaps = [[0,1],[2,3]]
    Output: 1

Example 2:
    Input: source = [1,2,3,4], target = [1,3,2,4], allowedSwaps = []
    Output: 2

Example 3:
    Input: source = [5,1,2,4,3], target = [1,5,4,2,3],
           allowedSwaps = [[0,4],[4,2],[1,3],[1,4]]
    Output: 0
"""

from collections import defaultdict
from typing import List


class UnionFind:
    def __init__(self, n: int):
        self.fa: list[int] = list(range(n))
        self.rank: list[int] = [0] * n

    def find(self, x: int) -> int:
        if self.fa[x] != x:
            self.fa[x] = self.find(self.fa[x])  # path compression
        return self.fa[x]

    def union(self, x: int, y: int):
        x = self.find(x)
        y = self.find(y)
        if x == y:
            return
        if self.rank[x] < self.rank[y]:
            x, y = y, x
        self.fa[y] = x
        if self.rank[x] == self.rank[y]:
            self.rank[x] += 1


class Solution:
    def minimumHammingDistance(
        self,
        source: List[int],
        target: List[int],
        allowedSwaps: List[List[int]],
    ) -> int:
        """
        Returns the minimum Hamming distance using Union-Find to group indices.

        Key insight: indices connected through allowedSwaps form a component
        within which source elements can be freely rearranged (swaps can be
        chained and repeated). The problem reduces to: for each component,
        how many target values cannot be matched by available source values?

        Algorithm:
        1. Build Union-Find: union every (a, b) pair from allowedSwaps.
        2. Group source values by component root into a frequency map.
        3. For each index i, check if target[i] is available in its component's
           frequency map. If yes, consume it (decrement). If no, add 1 to
           Hamming distance.

        Time complexity:  O(n + m·α(n)) where m = len(allowedSwaps) and
                          α is the inverse Ackermann function (near-constant).
        Space complexity: O(n)
        """
        n: int = len(source)
        uf = UnionFind(n)

        for a, b in allowedSwaps:
            uf.union(a, b)

        sets: dict = defaultdict(lambda: defaultdict(int))
        for i in range(n):
            f = uf.find(i)
            sets[f][source[i]] += 1

        ans: int = 0
        for i in range(n):
            f = uf.find(i)
            if sets[f][target[i]] > 0:
                sets[f][target[i]] -= 1
            else:
                ans += 1

        return ans

    def minimumHammingDistanceDFS(
        self,
        source: List[int],
        target: List[int],
        allowedSwaps: List[List[int]],
    ) -> int:
        """
        Returns the minimum Hamming distance using DFS to find connected components.

        Builds an adjacency list from allowedSwaps, then finds all connected
        components via DFS. Within each component, counts source and target value
        frequencies and greedily matches them: for each source value, the number
        of matches is min(source_freq[v], target_freq[v]). Unmatched positions
        contribute to Hamming distance.

        Time complexity:  O(n + m) where m = len(allowedSwaps).
        Space complexity: O(n + m) for the adjacency list and visited set.
        """
        n = len(source)
        G: list[list[int]] = [[] for _ in range(n)]
        for a, b in allowedSwaps:
            G[a].append(b)
            G[b].append(a)

        visited: set[int] = set()
        ans = 0

        def dfs(node: int, connected: set[int]) -> None:
            if node in connected:
                return
            connected.add(node)
            for child in G[node]:
                dfs(child, connected)

        for i in range(n):
            if i in visited:
                continue
            connected: set[int] = set()
            dfs(i, connected)

            counter_src: dict[int, int] = defaultdict(int)
            counter_tgt: dict[int, int] = defaultdict(int)
            for j in connected:
                counter_src[source[j]] += 1
                counter_tgt[target[j]] += 1

            unmatched = len(connected)
            for k, v in counter_src.items():
                unmatched -= min(v, counter_tgt[k])
            ans += unmatched
            visited |= connected

        return ans


if __name__ == "__main__":
    solution = Solution()

    test_cases = [
        ([1, 2, 3, 4], [2, 1, 4, 5], [[0, 1], [2, 3]], 1),
        ([1, 2, 3, 4], [1, 3, 2, 4], [],               2),
        ([5, 1, 2, 4, 3], [1, 5, 4, 2, 3], [[0, 4], [4, 2], [1, 3], [1, 4]], 0),
        ([1], [1], [], 0),
        ([1], [2], [], 1),
    ]

    for label, method in [
        ("minimumHammingDistance   ", solution.minimumHammingDistance),
        ("minimumHammingDistanceDFS", solution.minimumHammingDistanceDFS),
    ]:
        for source, target, swaps, expected in test_cases:
            result = method(source, target, swaps)
            assert result == expected, \
                f"{label} failed: source={source}, target={target}, swaps={swaps} → expected {expected}, got {result}"
        print(f"  {label} passed")

    print("All test cases passed!")
